### PR TITLE
Sh/archiver to jszip swap

### PR DESCRIPTION
**Pull request by shetzel**
### What does this PR do?
Replaces the archiver library with jszip.

### What issues does this PR fix or reference?
@W-13196750@

QA Notes:
Be sure to test using both NodeJS v18.16.0 and v18.15.0, but primarily 18.16.0.
1. conversion of static resources
2. deploys, both source and metadata format with static resources
3. retrieves
4. package version create with static resources

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -28,12 +28,12 @@
     "@salesforce/core": "^3.36.2",
     "@salesforce/kit": "^1.9.2",
     "@salesforce/ts-types": "^1.7.2",
-    "archiver": "^5.3.1",
```
Comment (mshanemc) [inline package.json:31]: be sure to take the types out of devDeps

File: src/client/metadataApiDeploy.ts [hunk 2]
```diff
@@ -437,20 +436,38 @@ export class MetadataApiDeploy extends MetadataTransfer<
...
-      // the typings on it are misleading and unintended.  More info https://github.com/archiverjs/node-archiver/issues/476
-      // If you await it, bad things happen, like the convert process exiting silently.  https://github.com/forcedotcom/cli/issues/1791
-      // leave the void as it is
-      void zip.finalize();
-      return stream2buffer(zip);
+
+      const zip = JSZip();
+
+      const zipDirRecursive = (dir: string): void => {
+        const list = fs.readdirSync(dir);
```
Comment (mshanemc) [inline src/client/metadataApiDeploy.ts:448]: `readdir` has an option for [`withFileTypes`](https://nodejs.org/api/fs.html#fspromisesreaddirpath-options) that will give you the isDirectory value without having to do _another_ fs operation inside the loop.

File: src/convert/streams.ts [hunk 3]
```diff
@@ -183,19 +183,17 @@ export class StandardWriter extends ComponentWriter {
...
-  private zip: Archiver = createArchive('zip', { zlib: { level: 3 } });
-  private buffers: Buffer[] = [];
+  private zip = JSZip();
+  private zipBuffer?: Buffer;
 
   public constructor(rootDestination?: SourcePath) {
     super(rootDestination);
-    void pipeline(this.zip, this.getOutputStream());
+    const destination = rootDestination ? `for: ${rootDestination}` : 'in memory';
+    this.logger.debug(`generating zip ${destination}`);
```
Comment (mshanemc) [inline src/convert/streams.ts:192]: if the logger is only used in ZipWriter, why not create it on the lowest level so all the other writers don't suffer the perf impact of unused loggers.

then, if someone else needs it, they could hoist it higher in the hierarchy

File: src/convert/streams.ts [hunk 4]
```diff
@@ -210,10 +208,7 @@ export class ZipWriter extends ComponentWriter {
           }
           // everything else can be zipped immediately to reduce the number of open files (windows has a low limit!) and help perf
           const streamAsBuffer = await stream2buffer(writeInfo.source);
-          return streamAsBuffer.length
-            ? this.addToZip(streamAsBuffer, writeInfo.output)
-            : // these will be zero-length files, which archiver supports via stream but not buffer
-              this.addToZip(writeInfo.source, writeInfo.output);
+          return this.addToZip(streamAsBuffer, writeInfo.output);
```
Comment (mshanemc) [inline src/convert/streams.ts:211]: cool!

**Review by mshanemc — CHANGES_REQUESTED** [review]

File: src/convert/streams.ts (hunk 3 above)
Comment (shetzel) [reply src/convert/streams.ts:192]: It's also used in `StandardWriter`

**Review by mshanemc — APPROVED** [review]
commits seem fine, no obvious perf impacts (looking especially at the `sourceToZip` across both OS).  

https://forcedotcom.github.io/source-deploy-retrieve/perf-Linux/
https://forcedotcom.github.io/source-deploy-retrieve/perf-Windows/

**Comment by mshanemc** [comment]
QA notes:
✅ deploy a static resource (single file) to org
✅ deploy a static resource (single file, but empty) to org

✅ deploy a zip containing one empty text file
✅  deploy an empty static resource folder as a zip (same error as current SDR/PDR combo)
| Error testBasic The specified Static Resource is not a valid zip file 

📓  Scenario: static resource has 1 zip file in it.  That file is forceIgnored.  [we got an issue about this scenario once](https://github.com/forcedotcom/source-deploy-retrieve/pull/663/files), and I saw the code change related to it.

current SDR/PDR throws this error **before sending anything to the API**.  
```Error (1): Metadata API request failed: force-app/main/default/staticresources/testBasic.resource-meta.xml: Expected source files for type 'StaticResource'```

👎🏻  this PR sends the bad zip file to the API, get a deployment failure, and shows this error.  I slightly prefer the old way (fail faster, especially if it's a large deployment)
```Error (1): Metadata API request failed: force-app/main/default/staticresources/testBasic.resource-meta.xml: Expected source files for type 'StaticResource'```

Still net improvement over archiver and happy to ship it.

**Pull request merged**